app/code/Magento/ConfigurableProductGraphQl/Model/Options/Metadata.php

```diff
@@ -0,0 +1,194 @@
+        $allowedAttributes = $this->configurableProductHelper->getAllowAttributes($product);
+        $attributeCodes = [];
+        foreach ($allowedAttributes as $attribute) {
+            $attributeCodes[$attribute->getAttributeId()] = $attribute->getProductAttribute()->getAttributeCode();
+        }
+        $availableSelections = [];
+        $availableSelectionProducts = [];
+        $variantData = null;
+        $addedAttributeOption = [];
+        foreach ($options['index'] as $productId => $availableProduct) {
+            if (empty($selectedOptions) || $this->hasProductRequiredOptions($selectedOptions, $availableProduct)) {
+                foreach ($availableProduct as $attributeId => $optionIndex) {
+                    $uid = $this->selectionUidFormatter->encode($attributeId, $optionIndex);
+                    if (!isset($addedAttributeOption[$attributeId])
+                        || !in_array($uid, $addedAttributeOption[$attributeId])
+                    ) {
+                        $addedAttributeOption[$attributeId][]
+                            = $availableSelections[$attributeId]['option_value_uids'][] = $uid;
+                        $availableSelections[$attributeId]['attribute_code']
+                            = $attributeCodes[$attributeId] ?? $this->getAttributeCode($attributeId);
+                        $availableSelectionProducts[] = $productId;
+                    }
+                }
+            }
+
+            if (!$variantData && $this->hasSelectionProduct($selectedOptions, $availableProduct)) {
+                $variantProduct = $this->productRepository->getById($productId);
+                $variantData = $variantProduct->getData();
+                $variantData['model'] = $variantProduct;
+            }
+        }
+        return [
+                'options_available_for_selection' => $availableSelections,
+                'variant' => $variantData,
+                'availableSelectionProducts' => array_unique($availableSelectionProducts),
+                'product' => $product
+        ];
+    }
+
+    /**
+     * Retrieve attribute code.
+     *
+     * @param $attributeId
+     * @return string
+     */
+    private function getAttributeCode($attributeId): string
+    {
+        $attribute = $this->attributeInterfaceFactory->create();
+        $this->attributeResuorce->load($attribute, $attributeId, AttributeInterface::ATTRIBUTE_ID);
+
+        return $attribute->getAttributeCode();
+    }
+
+    /**
+     * Get allowed products.
+     *
+     * @param ProductInterface $product
+     * @return ProductInterface[]
+     */
+    public function getAllowProducts(ProductInterface $product): array
+    {
+        $products = [];
+        $usedProducts = $product->getTypeInstance()->getUsedProducts($product, null);
+        foreach ($usedProducts as $usedProduct) {
+            if ((int) $usedProduct->getStatus() === Status::STATUS_ENABLED) {
```

Comment:
@cpartica @prabhuram93 Here I will add an additional check with isSalable on child product. But I have noticed that isSalable returns true if a child product is out of stock. 
I have tried to load product with ProductRepository and checked isSalable. In this case it returns false for out of stock product. 

The questions is why isSalable returns true if child products are loaded by collection? 

In term of performance, how is about a load by ProductRepository for child products?

Reply:
if we can have a collection where we load them all at once, using data loader then it's ok..but we can't load one at a time, it will be too slow
There's also this show products out of stock in catalog setting..